storage/innobase/handler/handler0alter.cc

```diff
@@ -1348,7 +1345,10 @@ innobase_need_rebuild(
 		return alter_options_need_rebuild(ha_alter_info, table);
 	}
 
-	return !!(ha_alter_info->handler_flags & INNOBASE_ALTER_REBUILD);
+	return !!(ha_alter_info->handler_flags & (
+			  INNOBASE_ALTER_REBUILD |
+			  ALTER_COLUMN_EQUAL_PACK_LENGTH |
+			  ALTER_COLUMN_UNVERSIONED));
```

Comment:
I believe that due to this, we will unnecessarily stop supporting VARCHAR extension and versioning removal for `ROW_FORMAT=COMPRESSED` tables. I tried the following test case:
```sql
--source include/have_innodb.inc
create table t1(c varchar(9)) engine=innodb row_format=compressed;
alter table t1 modify c varchar(10), algorithm=instant;
drop table t1;
```
Before this change, it would fail with
```
mysqltest: At line 4: query 'alter table t1 modify c varchar(10), algorithm=instant' failed: 1846: ALGORITHM=INSTANT is not supported. Reason: DROP INDEX. Try ALGORITHM=NOCOPY
```
After this change, even `ALGORITHM=NOCOPY` is refused, and the table would be rebuilt.

We might want to introduce a value of `alter_algorithm` that would avoid a conversion to a format for which `dict_table_t::is_instant()` would hold. If we ‘degrade’ all operations to rebuild and ‘upgrade’ to instant ALTER later, it could be more difficult to introduce such an option.

Reply:
I believe that if `ALTER_OPTIONS | ALTER_COLUMN_EQUAL_PACK_LENGTH` is set and `!alter_options_need_rebuild()`, then `innobase_need_rebuild()` would not return `true` as the logic of this patch would expect. (Example: `ALTER TABLE t1 page_compression_level=9, MODIFY COLUMN c VARCHAR(10);`)

Furthermore, I think that `export_vars.instant_alter_column` should only be incremented for those cases where a table rebuild has been avoided by converting to `dict_table_t::is_instant()` format. The patch below would be a step towards that direction:
```diff
diff --git a/storage/innobase/handler/handler0alter.cc b/storage/innobase/handler/handler0alter.cc
index d5642016733..cf24a4df8a9 100644
--- a/storage/innobase/handler/handler0alter.cc
+++ b/storage/innobase/handler/handler0alter.cc
@@ -1060,7 +1060,6 @@ struct ha_innobase_inplace_ctx : public inplace_alter_handler_ctx
 
 		instant_table = new_table;
 		new_table = old_table;
-		export_vars.innodb_instant_alter_column++;
 
 		instant_table->prepare_instant(*old_table, col_map,
 					       first_alter_pos);
@@ -1331,12 +1330,15 @@ static bool alter_options_need_rebuild(
 (or perform instant operation).
 @param[in] ha_alter_info	the ALTER TABLE operation
 @param[in] table		metadata before ALTER TABLE
+@param[in] really_need_rebuild	whether we really must rebuild the table
+				(or convert to dict_table_t::is_instant())
 @return whether it is necessary to rebuild the table or to alter columns */
 static MY_ATTRIBUTE((nonnull, warn_unused_result))
 bool
 innobase_need_rebuild(
 	const Alter_inplace_info*	ha_alter_info,
-	const TABLE*			table)
+	const TABLE*			table,
+	bool				really_need_rebuild = false)
 {
 	if ((ha_alter_info->handler_flags & ~(INNOBASE_INPLACE_IGNORE
 					      | INNOBASE_ALTER_NOREBUILD
@@ -1345,10 +1347,14 @@ innobase_need_rebuild(
 		return alter_options_need_rebuild(ha_alter_info, table);
 	}
 
-	return !!(ha_alter_info->handler_flags & (
-			  INNOBASE_ALTER_REBUILD |
-			  ALTER_COLUMN_EQUAL_PACK_LENGTH |
-			  ALTER_COLUMN_UNVERSIONED));
+	if (!really_need_rebuild) {
+		return !!(ha_alter_info->handler_flags
+			  & (INNOBASE_ALTER_REBUILD |
+			     ALTER_COLUMN_EQUAL_PACK_LENGTH |
+			     ALTER_COLUMN_UNVERSIONED));
+	}
+
+	return !!(ha_alter_info->handler_flags & INNOBASE_ALTER_REBUILD);
 }
 
 /** Check if virtual column in old and new table are in order, excluding
@@ -6532,6 +6538,10 @@ prepare_inplace_alter_table_dict(
 				= ctx->old_table->persistent_autoinc;
 		}
 
+		if (innobase_need_rebuild(ha_alter_info, old_table, true)) {
+			export_vars.innodb_instant_alter_column++;
+		}
+
 		ctx->prepare_instant();
 	}
 
```

Reply:
I updated your patch with that:
```diff
diff --git a/storage/innobase/handler/handler0alter.cc b/storage/innobase/handler/handler0alter.cc
index cf24a4df8a9..2508e99d6cc 100644
--- a/storage/innobase/handler/handler0alter.cc
+++ b/storage/innobase/handler/handler0alter.cc
@@ -2121,7 +2121,8 @@ ha_innobase::check_if_supported_inplace_alter(
        Field **af = altered_table->field;
        bool fts_need_rebuild = false;
        need_rebuild = need_rebuild
-               || innobase_need_rebuild(ha_alter_info, table);
+               || innobase_need_rebuild(ha_alter_info, table,
+                                        !m_prebuilt->table->supports_instant());
 
        while (Create_field* cf = cf_it++) {
                DBUG_ASSERT(cf->field
@@ -7935,9 +7936,13 @@ ha_innobase::prepare_inplace_alter_table(
        const ha_table_option_struct& alt_opt=
                *ha_alter_info->create_info->option_struct;
 
-       if (!(ha_alter_info->handler_flags & (INNOBASE_ALTER_DATA
-                                             | ALTER_COLUMN_EQUAL_PACK_LENGTH
-                                             | ALTER_COLUMN_UNVERSIONED))
+       alter_table_operations options = INNOBASE_ALTER_DATA;
+       if (m_prebuilt->table->supports_instant()) {
+               options |= ALTER_COLUMN_EQUAL_PACK_LENGTH;
+               options |= ALTER_COLUMN_UNVERSIONED;
+       }
+
+       if (!(ha_alter_info->handler_flags & options)
            || ((ha_alter_info->handler_flags & ~(INNOBASE_INPLACE_IGNORE
                                                  | INNOBASE_ALTER_NOCREATE
                                                  | INNOBASE_ALTER_INSTANT))
```

```sql
--source include/have_innodb.inc
create table t1(c varchar(9)) engine=innodb row_format=compressed;
alter table t1 modify c varchar(10), algorithm=instant;
drop table t1;
```

Now in that test case `innobase_instant_try()` must be called to update both data dictionary and cache. Instant inside nocopy.